plugins/modules/win_listen_ports_facts.ps1

```diff
@@ -0,0 +1,81 @@
+#!powershell
+
+# Copyright: (c) 2022, DataDope (@datadope-io)
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#AnsibleRequires -CSharpUtil Ansible.Basic
+
+$spec = @{
+    options = @{
+        date_format = @{ type = 'str'; default = '%c' }
+        tcp_filter = @{ type = 'list'; elements = 'str'; default = 'Listen' }
+    }
+    supports_check_mode = $true
+}
+
+$module = [Ansible.Basic.AnsibleModule]::Create($args, $spec)
+
+$date_format = $module.Params.date_format
+$tcp_filter = $module.Params.tcp_filter
+
+# Structure of the response the script will return
+$ansibleFacts = @{
+    tcp_listen = @()
+    udp_listen = @()
+}
+
+# Build an index of the processes based on the PID
+$processes = @{}
+Get-Process -IncludeUserName | ForEach-Object {
```

Comment:
Looks like PowerShell 3.0 does not have the `-IncludeUserName` parameter which is why CI is failing. There are 3 things I can see you do to solve this problem

* Conditionally add this parameter and `user` return value

```powershell
$getProcCommand = Get-Command -Name Get-Process
$getProcParams = @{}
if ('IncludeUserName' -in $getProcCommand.Parameters.Keys) {
    $getProcParams.IncludeUserName = $true
}
Get-Process @getProcParams | ForEach-Object { ... }
```

You also need to update `Build-Listener` to handle `UserName` not being present or just update the docs saying `user` will be null on hosts with PowerShell 3.0.

* Mark the module as needing PowerShell 4.0 or newer in the docs and skip 2012 in the tests `skip/windows/2012` in the aliases file

The docs would need to updated to reflect this.

* Use PInvoke to get the username from the process yourself

This is quite involved but luckily we have some helper utils to do this for you

```csharp
#AnsibleRequires -CSharpUtil Ansible.AccessToken

Function Get-ProcessUserName {
    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSAvoidUsingEmptyCatchBlock", "",
        Justification = "Getting the username is optional so errors are ignored")]
    [CmdletBinding()]
    param (
        [Parameter(Mandatory)]
        [int]
        $Id
    )

    try {
        $procHandle = [Ansible.Basic.TokenUtil]::OpenProcess($Id, "QueryInformation", $false)
        $token = [Ansible.Basic.TokenUtil]::OpenProcessToken($procHandle, "Query")
        $user = [Ansible.Basic.TokenUtil]::GetTokenUser($token)
        $user.Translate([System.Security.Principal.NTAccount]).Value
    }
    catch {}  # Most likely doesn't have enough rights, just silently ignore
}
```

Reply:
Thanks for the information @jborean93. I've followed a different approach to fix the missing `-IncludeUserName` functionality: replacing the `Get-Process` cmdlet with `Get-CimInstance -ClassName Win32_Process` and using `GetOwner`, so no compromise is needed (no conditional `user` returning, no PowerShell 4 requierement or not a possible silent error).

Reply:
It seems like Powershell 3 has no 'Bound' State for the TCP ports. I'll update the docs to specify it and the tests so 'Bound' is not one of the states used to test globally.